src/ttlv/fast_scan.rs

```diff
@@ -0,0 +1,1009 @@
+        let available = rest.as_flattened().len();
+
+        if usize::MAX as u32 == u32::MAX {
+            // 'usize' is at least as big as 'u32'.
+            available >= length as usize
+        } else {
+            // 'u32' is bigger than 'usize'.
+            available as u32 >= length
+        }
+    }
+
+    /// Finish the scanner.
+    ///
+    /// If there is remaining input, or an error was delayed, [`Err`] is
+    /// returned.  If additional remaining input is allowed, use
+    /// [`Self::finish_non_exhaustive()`] instead.
+    pub fn finish(self) -> Result<(), FastScanError> {
+        if self.is_empty() && !self.delayed_error() {
+            Ok(())
+        } else {
+            Err(FastScanError::assert())
+        }
+    }
+
+    /// Finish the scanner non-exhaustively.
+    ///
+    /// Additional input is ignored.  [`Err`] is returned only if an error is
+    /// delayed.  If additional remaining input is not allowed, use
+    /// [`Self::finish()`] instead.
+    pub fn finish_non_exhaustive(self) -> Result<(), FastScanError> {
+        if !self.delayed_error() {
+            Ok(())
+        } else {
+            Err(FastScanError::assert())
+        }
+    }
+}
+
+impl<'a> FastScanner<'a> {
+    /// Scan an expected structure.
+    ///
+    /// The returned sub-scanner can be used to scan the fields of the
+    /// `struct`; `self` will independently continue from the end of the
+    /// `struct`.
+    #[inline]
+    pub fn scan_struct(&mut self, tag: Tag) -> Result<Self, FastScanError> {
+        let expected_header = (tag.value() as u64) << 40 | 0x01_00000000;
+        let [header, rest @ ..] = self.remaining else {
+            return Err(FastScanError::assert());
+        };
+        let header = u64::from_be_bytes(*header);
+        self.failure_word |= (expected_header ^ header) & 0xFFFFFFFF_00000007;
+        let length = header as u32 as usize / 8;
+        let (body, rest) = rest.split_at_checked(length).ok_or(FastScanError::assert())?;
+        self.remaining = rest;
+
+        Ok(Self::from_blocks(body))
+    }
+
+    /// Scan an expected structure with a pre-determined length.
+    ///
+    /// The returned sub-scanner can be used scan the fields of the `struct`;
+    /// `self` will independently continue from the end of the `struct`.  If
+    /// the length encoded in the input is different from the pre-determined
+    /// length, the error will be delayed and both scanners will return
+    /// garbage.
+    #[inline]
+    pub fn scan_fixed_struct(&mut self, tag: Tag, length: u32) -> Result<Self, FastScanError> {
+        debug_assert!(length.is_multiple_of(8));
+        let expected_header = ((tag.value() as u64) << 40 | 0x01 << 32 | length as u64).to_be();
+        if self.remaining.len() < 1 + length as usize / 8 {
+            return Err(FastScanError::assert());
+        }
+        let (header, rest) = self.remaining.split_first().unwrap();
+        let header = u64::from_ne_bytes(*header);
```

Comment:
I see `from_be_bytes()` above but `from_ne_bytes()` below. Should they all be the same?